src/consensus/aft/raft.h

```diff
@@ -1103,43 +1106,55 @@ namespace aft
           return;
         }
 
-        state->last_idx = i;
-
-        Term sig_term = 0;
-        Index sig_index = 0;
-        auto tx = store->create_tx();
-        kv::DeserialiseSuccess deserialise_success;
-        ccf::PrimarySignature sig;
-        kv::ConsensusHookPtrs hooks;
-        if (consensus_type == ConsensusType::BFT)
+        auto ds = store->apply(entry, consensus_type, public_only);
+        if (ds == nullptr)
         {
-          deserialise_success = store->deserialise_views(
-            entry, hooks, public_only, &sig_term, &sig_index, &tx, &sig);
+          LOG_DEBUG_FMT("failed to deserialize we failed to apply");
+          send_append_entries_response(
+            r.from_node, AppendEntriesResponseType::FAIL);
+          return;
         }
-        else
+        append_entries.push_back(std::make_tuple(std::move(ds), i));
+      }
+
+      for (auto& ae : append_entries)
+      {
+        auto& [ds, i] = ae;
+        state->last_idx = i;
+
+        kv::ApplySuccess apply_success = kv::ApplySuccess::FAILED;
+        if (ds != nullptr)
         {
-          deserialise_success =
-            store->deserialise(entry, hooks, public_only, &sig_term);
+          apply_success = ds->execute();
+          if (apply_success == kv::ApplySuccess::FAILED)
+          {
+            state->last_idx = i - 1;
+            ledger->truncate(state->last_idx);
+            send_append_entries_response(
+              r.from_node, AppendEntriesResponseType::FAIL);
+            return;
+          }
         }
 
-        for (auto& hook : hooks)
+        for (auto& hook : ds->get_hooks())
```

Comment:
Will this segfault if `ds` is `nullptr`?

Reply:
ds cannot be nullptr because we check this on line 1110. I am removing the check on 1126 at it is not needed